File: authserver/src/main/java/l2s/authserver/network/gamecomm/gs2as/AuthRequest.java
package l2s.authserver.network.gamecomm.gs2as;

import io.netty.buffer.ByteBuf;
import l2s.authserver.GameServerManager;
import l2s.authserver.network.gamecomm.GameServerDescription;
import l2s.authserver.network.gamecomm.GameServerDescription.HostInfo;
import l2s.authserver.network.gamecomm.ReceivablePacket;
import l2s.authserver.network.gamecomm.as2gs.AuthResponse;
import l2s.authserver.network.gamecomm.as2gs.LoginServerFail;
import l2s.authserver.network.gamecomm.vertx.GameServerConnection;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

public class AuthRequest extends ReceivablePacket {
    private static final Logger _log = LoggerFactory.getLogger(AuthRequest.class);
    private final int _protocolVersion;
    private final List<HostInfo> _hosts;
    private final int _serverType;
    private final int _ageLimit;
    private final boolean _gmOnly;
    private final boolean _brackets;
    private final boolean _pvp;
    private final int _maxOnline;

    public AuthRequest(GameServerConnection client, ByteBuf byteBuf) {
        super(client, byteBuf);

        _protocolVersion = byteBuf.readInt();
        byteBuf.readUnsignedByte();
        byteBuf.readUnsignedByte();
        _serverType = byteBuf.readInt();
        _ageLimit = byteBuf.readInt();
        _gmOnly = byteBuf.readUnsignedByte() == 1;
        _brackets = byteBuf.readUnsignedByte() == 1;
        _pvp = byteBuf.readUnsignedByte() == 1;
        readString(byteBuf);
        readString(byteBuf);
        for (int ports = byteBuf.readUnsignedShort(), i = 0; i < ports; ++i)
            byteBuf.readUnsignedShort();
        _maxOnline = byteBuf.readInt();
        int hostsCount = byteBuf.readUnsignedByte();
        _hosts = new ArrayList<>(hostsCount);
        for (int j = 0; j < hostsCount; ++j) {
            int id = byteBuf.readUnsignedByte();
            String ip = readString(byteBuf);
            String innerIP = readString(byteBuf);
            int port = byteBuf.readUnsignedShort();
            String key = readString(byteBuf);
            _hosts.add(new HostInfo(id, ip, innerIP, port, key));
        }
    }

    @Override
    protected void runImpl(GameServerConnection client) {
        _log.info("Trying to register gameserver: IP[" + client.getRemoteAddress() + "]");
        GameServerDescription gs = client.getGameServerDescription();
        for (HostInfo host : _hosts)
            if (GameServerManager.getInstance().registerGameServer(host.getId(), client))
                gs.addHost(host);
            else {
                client.sendPacket(new LoginServerFail("Gameserver registration on ID[" + host.getId() + "] failed. ID[" + host.getId() + "] is already in use!", false));
                client.sendPacket(new LoginServerFail("Free ID[" + host.getId() + "] or change to another ID, and restart your authserver or gameserver!", false));
            }
        if (gs.getHosts().length > 0) {
            gs.setProtocol(_protocolVersion);
            gs.setServerType(_serverType);
            gs.setAgeLimit(_ageLimit);
            gs.setGmOnly(_gmOnly);
            gs.setShowingBrackets(_brackets);
            gs.setPvp(_pvp);
            gs.setMaxPlayers(_maxOnline);
            client.setAuthed(true);
            _log.info("Gameserver registration successful.");
            client.sendPacket(new AuthResponse(gs));
            return;
        }
        client.sendPacket(new LoginServerFail("Gameserver registration failed. All ID's is already in use!", true));
        _log.info("Gameserver registration failed.");
    }

}

File: authserver/src/main/java/l2s/authserver/network/gamecomm/gs2as/AuthRequest.kt
package l2s.authserver.network.gamecomm.gs2as

import io.netty.buffer.ByteBuf
import l2s.authserver.AuthServer
import l2s.authserver.GameServerManager
import l2s.authserver.network.gamecomm.ReceivablePacket
import l2s.authserver.network.gamecomm.as2gs.AuthResponse
import l2s.authserver.network.gamecomm.as2gs.LoginServerFail
import l2s.authserver.network.gamecomm.vertx.GameServerConnection
import l2s.commons.net.HostInfo
import org.slf4j.Logger
import org.slf4j.LoggerFactory


/**
 * @reworked by Bonux
 */
class AuthRequest(client: GameServerConnection, byteBuf: ByteBuf) : ReceivablePacket(client, byteBuf) {

    private val _protocolVersion: Int
    private val _hosts: Array<HostInfo>
    private val _serverType: Int
    private val _ageLimit: Int
    private val _gmOnly: Boolean
    private val _brackets: Boolean
    private val _pvp: Boolean
    private val _maxOnline: Int

    init {
        _protocolVersion = byteBuf.readInt()
        require(_protocolVersion == AuthServer.AUTH_SERVER_PROTOCOL) {
            "Auth server protocol version is wrong: $_protocolVersion."
        }
        _serverType = byteBuf.readInt()
        _ageLimit = byteBuf.readInt()
        _gmOnly = byteBuf.readUnsignedByte() == 1.toShort()
        _brackets = byteBuf.readUnsignedByte() == 1.toShort()
        _pvp = byteBuf.readUnsignedByte() == 1.toShort()
        _maxOnline = byteBuf.readInt()
        val hostsCount: Int = byteBuf.readUnsignedByte().toInt()
        _hosts = Array(hostsCount) {
            val id: Int = byteBuf.readUnsignedByte().toInt()
            val address: String = readString(byteBuf)
            val port: Int = byteBuf.readUnsignedShort()
            val key: String = readString(byteBuf)
            val maskCount: Int = byteBuf.readUnsignedByte().toInt()
            val host = HostInfo(id, address, port, key)
            for (m in 0 until maskCount) {
                val subAddress: String = readString(byteBuf)
                val subnetAddress = ByteArray(byteBuf.readInt())
                byteBuf.readBytes(subnetAddress)
                val subnetMask = ByteArray(byteBuf.readInt())
                byteBuf.readBytes(subnetMask)
                host.addSubnet(subAddress, subnetAddress, subnetMask)
            }

            host
        }
    }

    override fun runImpl(client: GameServerConnection) {
        if (_protocolVersion != AuthServer.AUTH_SERVER_PROTOCOL) {
            logger.warn("Authserver and gameserver have different versions! Please update your servers.")
            client.sendPacket(
                LoginServerFail(
                    "Authserver and gameserver have different versions! Please update your servers.",
                    false
                )
            )
            return
        }

        logger.info("Trying to register gameserver: IP[" + client.remoteAddress + "]")
        for (host in _hosts) {
            val registerResult =
                GameServerManager.getInstance().registerGameServer(host, client)
            if (registerResult == GameServerManager.SUCCESS_GS_REGISTER)
                client.description.addHost(host)
            else {
                when (registerResult) {
                    GameServerManager.FAIL_GS_REGISTER_DIFF_KEYS -> {
                        client.sendPacket(
                            LoginServerFail(
                                "Gameserver registration on ID[" + host.id.toString() + "] failed. Registered different keys!",
                                false
                            )
                        )
                        client.sendPacket(
                            LoginServerFail(
                                "Set the same keys in authserver and gameserver, and restart them!",
                                false
                            )
                        )
                    }
                    GameServerManager.FAIL_GS_REGISTER_ID_ALREADY_USE -> {
                        client.sendPacket(
                            LoginServerFail(
                                "Gameserver registration on ID[" + host.id.toString() + "] failed. ID[" + host.id.toString() + "] is already in use!",
                                false
                            )
                        )
                        client.sendPacket(
                            LoginServerFail(
                                "Free ID[" + host.id.toString() + "] or change to another ID, and restart your authserver or gameserver!",
                                false
                            )
                        )
                    }
                    GameServerManager.FAIL_GS_REGISTER_ERROR -> {
                        client.sendPacket(
                            LoginServerFail(
                                "Gameserver registration on ID[" + host.id.toString() + "] failed. You have some errors!",
                                false
                            )
                        )
                        client.sendPacket(LoginServerFail("To solve the problem, contact the developer!", false))
                    }
                }
            }
        }

        val description = client.description
        if (description.hosts.isNotEmpty()) {
            description.protocol = _protocolVersion
            description.serverType = _serverType
            description.ageLimit = _ageLimit
            description.isGmOnly = _gmOnly
            description.isShowingBrackets = _brackets
            description.isPvp = _pvp
            description.maxPlayers = _maxOnline
            description.store()
            client.isAuthed = true
        } else {
            client.sendPacket(LoginServerFail("Gameserver registration failed. All ID's is already in use!", true))
            logger.info("Gameserver registration failed.")
            return
        }
        logger.info("Gameserver registration successful.")
        client.sendPacket(AuthResponse(description))
    }

    companion object {

        private val logger: Logger = LoggerFactory.getLogger(AuthRequest::class.java)

    }

}